sql/row: avoid NegotiateAndSend on a multi-batch request

otan opened the pull request:
Avoid the double NegotiateAndSend on a bounded staleness request if
there are multiple column families for the given row by tracking
whether a negotiate was already made. If so, use Send, as
SetFixedTimestamp would already have the correct value.

Release justification: fix to new feature

Resolves: #69063

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69350)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69350#-:-MhxjmNQ82Pifobum3OA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/row/kv_fetcher.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69350#-MhxjALX4KAYlt6lkB_6:-MhxjALYDyswqLo4h2Px:bd14fp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/05125ef88e09a2e5d6313356587792a6cc22efd7/pkg/sql/row/kv_fetcher.go#L68)):*
> ```Go
> 		sendFn = func(ctx context.Context, ba roachpb.BatchRequest) (br *roachpb.BatchResponse, _ error) {
> 			ba.RoutingPolicy = roachpb.RoutingPolicy_NEAREST
> 			ba.BoundedStaleness = bsHeader
> ```

We don't want this, right? Won't this hit the assertion we just added on the Send path? Do we need a test that intentionally hits the Send path?
___
*[pkg/sql/row/kv_fetcher.go, line 75 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69350#-MhxjSu75vK8SSyDr8zf:-MhxjSu8A7Ddc0wZnhLD:b3isbuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/05125ef88e09a2e5d6313356587792a6cc22efd7/pkg/sql/row/kv_fetcher.go#L75)):*
> ```Go
> 			if !negotiated {
> 				ba.BoundedStaleness = bsHeader
> 				br, pErr = txn.NegotiateAndSend(ctx, ba)
> ```

Do you mind opening an issue for me to add a test to `txn_external_test.go` that ensures that a bounded staleness read that returns a resume span still negotiates correctly and fixes the transaction timestamp?


<!-- Sent from Reviewable.io -->

otan committed:
sql/row: avoid NegotiateAndSend on a multi-batch request

Avoid the double NegotiateAndSend on a bounded staleness request if
there are multiple column families for the given row by tracking
whether a negotiate was already made. If so, use Send, as
SetFixedTimestamp would already have the correct value.

Release justification: fix to new feature

Release note: None

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69350)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/sql/row/kv_fetcher.go, line 68 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69350#-MhxjALX4KAYlt6lkB_6:-Mhz2gfsBg_5BeuWLbik:b-fxkyf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/05125ef88e09a2e5d6313356587792a6cc22efd7/pkg/sql/row/kv_fetcher.go#L68)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We don't want this, right? Won't this hit the assertion we just added on the Send path? Do we need a test that intentionally hits the Send path?
</blockquote></details>

urgh! i rebased this wrong. my bad. this test would have failed (...flakily / under stress...)


<!-- Sent from Reviewable.io -->
